lib/Onyx.js

```diff
@@ -634,7 +636,7 @@ function sendDataToConnection(mapping, val, matchedKey) {
         return;
     }
 
-    if (_.isFunction(mapping.callback)) {
+    if (_.isFunction(mapping.callback) && (isCollectionKey(mapping.key) || matchedKey)) {
```

Comment:
This change prevents extra calls to `Onyx.connect` subscribers with value `null` and key `undefined`. `withOnyx` waits for all connection callbacks to have been called once before rendering, so that's why we added this type of callback. However, I don't think that this type of callback is helpful to other library code using `Onyx.connect` directly, so I prevented that here. `withOnyx` callbacks are handled in the block above which terminates with an early return.

Reply:
Can you clarify which call to `sendDataToConnection()` we are trying to prevent and why? There are only a few of them - so I worry that we are adding this conditional exception in the wrong place. And maybe don't really need this change at all.

Reply:
Yes, if you do this:

```js
const myCallback = () => {};
Onyx.connect({
   key: 'someKey',
   callback: myCallback,
});
```

then `myCallback` will _always_ first be called with `null, undefined` for the value and key. I think this is counter-intuitive and unhelpful in most cases.

The reason we do that is because `withOnyx` waits for each `Onyx.connect` callback to be called at least once before rendering the child component. So even if the data is missing, we call it with `null, undefined`.

My thinking is that by doing this we're exposing an internal implementation detail of `withOnyx` unnecessarily, and in the general case `Onyx.connect` should only execute the callback when there's some data in the store with the given key.

Reply:
Sounds like the main premise in your argument is that "no Onyx.connect() subscribers need this". 

**thought:** passing `null` also provides a way for an Onyx.connect() subscriber to see that it initialized _without a value from storage_. 

Is it helpful to know this? Do we have existing code that depends on this information? I think it's safer to assume we do than don't, but also don't want to audit the 446 usages 😄